sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+
+Applying the `indirectly_callable` property to a SYCL Kernel function is illegal
+and an implementation should produce a diagnostic for that.
+
+Applying the `indirectly_callable` property to an arbitrary device function,
+which is not a virtual member function has no effect.
+
+NOTE: This behavior may be changed in either future version of this extension or
+in another extensions.
+
+Virtual member functions that are decorated with the `indirectly_callable`
+property are considered to be device functions, i.e. they  must obey the
+restrictions listed in section 5.4 of the core SYCL specification "Language
+restrictions for device functions". Virtual member functions that are not
+decorated with this attribute do not need to obey these restrictions, even if
+other definitions of that virtual member function in other classes in the
+inheritance hierarchy are decorated with the attribute.
+
+[source,dpcpp]
+----
+using syclext = sycl::ext::oneapi::experimental;
+
+struct set_A;
+struct set_B;
+
+class Foo {
+public:
+  // properties to functions should be applied using the macro:
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void foo() {}
+
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void bar() {};
+
+  // both declaration and definition should be annotated:
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_B) void baz();
+};
+
+// both declaration and definition should be annotated:
+void SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_B) Foo::baz() {}
+
+// kernel calling virtual function should also be annotated:
+/* */.single_task(syclext::properties{syclext::calls_indirectly<set_A>}, [=] {
+  Foo *ptr = /* ... */;
+  ptr->bar()
+
+  // Note: this kernel can only call 'Foo::foo' and 'Foo::bar' but not
+  // 'Foo::baz', because the latter is declared within a different set.
+});
+----
+
+Reasons about why it may be necessary to put different virtual functions into
+different sets are explained in further sections, but for simplicity purposes
+both properties can be used without explicitly specifying a set, thus using
+the default set:
+
+[source,dpcpp]
+----
+using syclext = sycl::ext::oneapi::experimental;
+
+struct set_A;
+
+class Foo {
+public:
+  // This virtual member function belongs to the default set of virtual
+  // functions.
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {}
+
+  virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(set_A) void bar() {}
+};
+
+/* */.single_task(syclext::properties{syclext::calls_indirectly<>}, [=] {
+  Foo *ptr = /* ... */;
+  ptr->bar()
+
+  // Note: this kernel can only call 'Foo::foo' but not 'Foo::bar', because the
+  // latter belongs to a different (non-default) set of virtual functions.
+});
+----
+
+NOTE: `void` typename is explicitly reserved as identifier of the default
+virtual functions set. It can still be used, but with the expectation that it
+is also implicitly used by the extension implementation.
```

Comment:
This wording seems a little weird to me because "implicitly used by the extension implementation" makes me think that using the type `void` might somehow interfere with the implementation.  Can we say this instead?

> By convention, the type `void` is used to denote the default set of virtual functions.  When the typename is omitted from the `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` macro, the type `void` is used by default.  Applications may also explicitly use the type `void` to denote this default set of virtual functions.

Reply:
Thanks! I've applied the suggestion in a07048ffca79dec10df6257b2e7d396b4960c7b7